src/tensor_operations/matrix_decomposition.jl

```diff
@@ -369,46 +369,168 @@ function noinds_error_message(decomp::String)
    treating the ITensor as a matrix from the primed to the unprimed indices."
 end
 
+function add_trivial_index(A::ITensor, Ainds)
+  α = trivial_index(Ainds) #If Ainds[1] has no QNs makes Index(1), otherwise Index(QN()=>1)
+  vα = onehot(eltype(A), α => 1)
+  A *= vα
+  return A, vα, [α]
+end
+
+function add_trivial_index(A::ITensor, Linds, Rinds)
+  vαl, vαr = nothing, nothing
+  if isempty(Linds)
+    A, vαl, Linds = add_trivial_index(A, Rinds)
+  end
+  if isempty(Rinds)
+    A, vαr, Rinds = add_trivial_index(A, Linds)
+  end
+  return A, vαl, vαr, Linds, Rinds
+end
+
+remove_trivial_index(Q::ITensor, R::ITensor, vαl, vαr) = (Q * dag(vαl), R * dag(vαr))
+remove_trivial_index(Q::ITensor, R::ITensor, ::Nothing, vαr) = (Q, R * dag(vαr))
+remove_trivial_index(Q::ITensor, R::ITensor, vαl, ::Nothing) = (Q * dag(vαl), R)
+remove_trivial_index(Q::ITensor, R::ITensor, ::Nothing, ::Nothing) = (Q, R)
+
+#Force users to knowingly ask for zero indices using qr(A,()) syntax
 qr(A::ITensor; kwargs...) = error(noinds_error_message("qr"))
+rq(A::ITensor; kwargs...) = error(noinds_error_message("rq"))
+lq(A::ITensor; kwargs...) = error(noinds_error_message("lq"))
+ql(A::ITensor; kwargs...) = error(noinds_error_message("ql"))
 
-# TODO: write this in terms of combiners and then
-# call qr on the order-2 tensors directly
+# qr is exported by the LinearAlgebra module so we need acknowledge that to avoid
+# intermitent run time errors.
 function qr(A::ITensor, Linds...; kwargs...)
-  tags::TagSet = get(kwargs, :tags, "Link,qr")
+  qtag::TagSet = get(kwargs, :tags, "Link,qr") #tag for new index between Q and R
   Lis = commoninds(A, indices(Linds...))
   Ris = uniqueinds(A, Lis)
-
-  Lis_original = Lis
-  Ris_original = Ris
-  if isempty(Lis_original)
-    α = trivial_index(Ris)
-    vLα = onehot(eltype(A), α => 1)
-    A *= vLα
-    Lis = [α]
-  end
-  if isempty(Ris_original)
-    α = trivial_index(Lis)
-    vRα = onehot(eltype(A), α => 1)
-    A *= vRα
-    Ris = [α]
+  # Make a dummy index with dim=1 and incorporate into A so the Lis & Ris can never
+  # be empty.  A essentially becomes 1D after collection.
+  A, vαl, vαr, Lis, Ris = add_trivial_index(A, Lis, Ris)
+
+  #
+  #  Use combiners to render A down to a rank 2 tensor ready matrix QR routine.
+  #
+  CL, CR = combiner(Lis...), combiner(Ris...)
+  cL, cR = combinedind(CL), combinedind(CR)
+  AC = A * CR * CL
+  #
+  #  Make sure we don't accidentally pass the transpose into the matrix qr routine.
+  #
+  AC = permute(AC, cL, cR; allow_alias=true)
+  # qr the matrix.
+  QT, RT = qr(tensor(AC); kwargs...)
+  #
+  # correct the fluxes of the two tensors, such that Q has 0 flux for all blocks
+  # and R has the total flux of the system
+  # 
+  if hasqns(AC)
+    for b in nzblocks(QT)
+      i1 = inds(QT)[1]
+      i2 = inds(QT)[2]
+      r1 = inds(RT)[1]
+      newqn = -dir(i2) * flux(i1 => Block(b[1]))
+      ITensors.setblockqn!(i2, newqn, b[2])
+      ITensors.setblockqn!(r1, newqn, b[2])
+    end
```

Comment:
In principle this code shouldn't be necessary if the indices are built correctly in NDTensors, but that should not require using ITensors functions like `setblockqn!`. For example, Miles fixed a similar issue with the SVD code a while ago, from within NDTensors: https://github.com/ITensor/ITensors.jl/commit/c12e4f21187524eb18949469a6a4a496c350ca73. Can we do something similar here?

Reply:
Looks like you found a nice fix for this, that's great, thanks for investigating!

Reply:
Most welcome.  Much more elegant to build the new index at the NDTensors level.  There are two remaining problems for the case where R is order 2:
1) The indices of R should be ordered so that dense(R) displays an upper triangular matrix. (easy to fix)
2) RT returned from NDTensors is triangular, but R=RT*dag(CR) is often not triangular any more.  This is a consequence of the QN space combining and sorting done inside the combiner.  I don't think this can be fixed by building the index between Q&R in a different order because the external index of R (right index of T) could have split blocks, but the data passed to NDTensors has merged blocks.  Anyway that is my current understanding, but as you know my understandings in this area are often wrong!

I plan to add some code to make a simple gauge transform (after the R=RT*dag(CR) steps) to apply to Q and R in order to restore triangularity [word?].  I suppose we could make this optional, or provide a flag to turn it off for users who don't care.

Let me know if you think the gauge transform is the wrong approach,

Reply:
Before you get started on that, let's discuss it. I don't think either of those are necessary, since the index ordering of ITensors are internal details, and so is the ordering of the QN spaces after combining. Also, I'm not quite sure what you mean by your point 1., since if `R` is upper triangular, `dense(R)` should be as well since it should have the same exact elements (just with the block structure removed).

But in general, for ITensor `R` we should only guarantee that it is upper triangular _if_ it is permuted such that the index shared with `Q` is the first one (for a QR decomposition, it should be the last one for an RQ decomposition). But we don't guarantee which order of indices it is output with, so that we could choose the index ordering to avoid unnecessary permutations.

Reply:
Perhaps I'm just misunderstanding your point though, could you share a minimal example of what you think is incorrect about the current implementation?

Reply:
I agree point #1 is minor, it means a new user will be surprised to sometimes see what *appears* be lower triangular when they do @show dense(R).  An experienced user will understand that index ordering is arbitrary and not be concerned.

I will make a small example showing problem #2.

Reply:
Right, but we shouldn't fix 1. in the code, if ordering the indices in the "counterintuitive" way is more performant.

Reply:
OK here is a small example:

```
using ITensors
#QN ordering we might see in an MPO
qns=[QN("Sz",0)=>1,QN("Sz",-1)=>1,QN("Sz",1)=>1,QN("Sz",0)=>1]
i=Index(qns,"i")
A=randomITensor(i,i')
Q,R,q=qr(A,i)
@show dense(R)
#QN ordering carefully chosen to not get re-ordered by combiners
qns=[QN("Sz",-1)=>1,QN("Sz",0)=>1,QN("Sz",0)=>1,QN("Sz",1)=>1]
i=Index(qns,"i")
A=randomITensor(i,i')
Q,R,q=qr(A,i)
@show dense(R)

```
and the output:

> dense(R) = ITensor ord=2
> Dim 1: (dim=4|id=257|"i")'
> Dim 2: (dim=4|id=826|"Link,qr")
> Dense{Float64, Vector{Float64}}
>  4×4
>  0.000   1.677   0.000  0.000
>  0.373   0.000   0.000  0.000
>  0.000   0.000   0.000  0.055
>  0.000  -1.615  -0.215  0.000
> dense(R) = ITensor ord=2
> Dim 1: (dim=4|id=773|"i")'
> Dim 2: (dim=4|id=680|"Link,qr")
> Dense{Float64, Vector{Float64}}
>  4×4
>  1.607   0.000   0.000   0.000
>  0.000  -1.093   0.000   0.000
>  0.000   0.303  -1.127   0.000
>  0.000   0.000   0.000  -0.215
> ITensor ord=2 (dim=4|id=773|"i")' (dim=4|id=680|"Link,qr")
> Dense{Float64, Vector{Float64}}

Reply:
Thanks @JanReimers, I understand what you mean now. I think there is a more fundamental question here about what it means for an ITensor to be upper triangular, since as you've identified that is very basis-dependent.

For example, if you take a QR decomposition of an ITensor with multiple indices that end up an `R`, how do you define if it is upper triangular? You have to define a basis ordering for those multiple indices. I would argue that the way to "define" if it is upper triangular would be if it is upper triangular after you combine the indices that are unique to `R`, for example:
```julia
using ITensors

space = [QN("Sz",0)=>1, QN("Sz",-1)=>1, QN("Sz",1)=>1, QN("Sz",0)=>1]
i, j, k = Index.((space, space, space), ("i", "j", "k"))
A = randomITensor(i, j, k)
Q, R = qr(A, i)
q = commonind(Q, R)
C = combiner(j, k)
c = combinedind(C)
Matrix(R * C, q, c)
```
which outputs:
```julia
4×16 Matrix{Float64}:
 0.0  0.118356  0.345319  -0.394825  -0.507585   0.0      …   0.0       0.0       0.0       0.0       0.0       0.0
 0.0  0.0       0.0        0.0        0.0       -1.25421      1.62696   0.0       0.0       0.0       0.0       0.0
 0.0  0.0       0.0        0.0        0.0        0.0         -0.825069  0.0       0.0       0.0       0.0       0.0
 0.0  0.0       0.0        0.0        0.0        0.0          0.0       0.700738  0.219431  0.454174  0.329561  0.0
```
i.e. it is only well defined once you matricize it in the correct way, using the index ordering and index combining that the (matrix) QR decomposition used.

For the case where the ITensor is order 2 and the QR is treated as a matrix between the two indices, we _could_ special case that to ouput an upper triangular `R`, but I still think that we shouldn't force that and keep the same convention, that it is only upper triangular after using a combiner on `R` (which should give a deterministic way to order the basis in the same way that the matrix QR used internally).

Reply:
This leads to another point, which is that in general really people should specify both the left and right indices of the QR to make sure they know the correct index/basis ordering and therefore combine the indices correctly to get an upper triangular `R`. However, I find that this doesn't work:
```julia
using ITensors

space = [QN("Sz",0)=>1, QN("Sz",-1)=>1, QN("Sz",1)=>1, QN("Sz",0)=>1]
i, j, k = Index.((space, space, space), ("i", "j", "k"))
A = randomITensor(i, j, k)
Q, R = qr(A, (i,), (j, k))
q = commonind(Q, R)
C = combiner(j, k)
c = combinedind(C)
Matrix(R * C, q, c)
```
since it puts all of the input indices `(i, j, k)` onto `Q`. I think you need to add a version:
```julia
qr(A::ITensor, Linds, Rinds; kwargs...)
```
like we have for `eigen` where you can specify both the left and right indices. That should be the "core" function, and then we can have some convenience functions:
```julia
qr(A::ITensor, Linds::Index...; kwargs...)
qr(A::ITensor, Linds; kwargs...)
```
which call the version `qr(A::ITensor, Linds, Rinds; kwargs...)` by determining `Rinds` from the set difference of the indices of `A` and `Linds`. That could be for the case where a user doesn't care about the basis ordering, for example if they just want the orthogonal tensor `Q`.

Reply:
> I don't think there will be much performance effects for fixing item #1, it is just one line in the outer routine:
> 
> ```
>  #
>  # Make sure that q is the first index of R so that dense(R) appears as upper triangular.
>  #
>  R = permute(R, q, noncommoninds(R, q)...; allow_alias=true) #make sure the q index is first.
> ```
> 
> But I think most important is to resolve item #2, where experienced users will be scratching their heads trying figure out what to do with a non-triangular R as the example above demonstrates.

As I said, I don't think this is something that needs to be "fixed" from our end, since we are allowed to output ITensors in any order we want, since mathematically the index ordering doesn't matter. So we should do extra work just to make printing nicer. Users should permute the tensors themselves if that is the ordering they want.

Reply:
Yes my code inside qr() checks 
`if hasqns(q) && order(R)==2
`
wrapping the gauge transform which is two red flags right there!  I will have to document the basis dependence well, so people understand.

I am pretty sure that for MPO compression I need to gauge fix to get R back to triangular, in order to maintain Parker's regular form.  i.e. those identity operators in the corners, really need to stay in the corners.  There may be some generalization where in they can move around, but that might be much more difficult to understand and program.  Is this a one off situation?  I don't know.  Should we provide a separate interface that users can call after qr, in order to restore triangularity if users think they need it?  Or more friendly a kwargs flag that does it inside qr, and errors 
`if flag==true && order(R)!=2?
`

In any case we can definitely support 
`qr(A::ITensor, Linds, Rinds; kwargs...)`
as the core function.  That will be easy.

Reply:
OK no problem.  So I think the remaining issues are:

1. Use qr(A::ITensor, Linds, Rinds; kwargs...) interface as the core function.
2. See if can we can reduce lines of code in NDTensors by implementing {qr,ql} instead of {qr,rq}.  If it looks promising it might take a while. 

Does that sound correct?

Reply:
Yeah, that's a good summary, thanks!